pkg/config/client/client_config.go

```diff
@@ -101,10 +103,23 @@ func (c ClientTransportOverrides) DefaultClientTransport(rt http.RoundTripper) h
 		return rt
 	}
 
-	transport.DialContext = (&net.Dialer{
+	dialer := &net.Dialer{
 		Timeout:   30 * time.Second,
 		KeepAlive: 30 * time.Second,
-	}).DialContext
+		Control: func(network, address string, c syscall.RawConn) error {
+			// Supported only on Linux
+			// TODO: provide code that will be platform depended
+			err := c.Control(func(fd uintptr) {
+				err := syscall.SetsockoptInt(int(fd), syscall.IPPROTO_TCP, /*TCP_USER_TIMEOUT taken from https://godoc.org/golang.org/x/sys/unix*/ 0x12, int(10 * time.Second/time.Millisecond))
```

Comment:
not sure what's the best value here

Reply:
According to https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die/, this option will short-circuit keepalive TCP_KEEPCNT * TCP_KEEPINTVL. Given that client-go sets the KEEPINTVL to 30 seconds, this means that we will send exactly one keepalive probe before this kicks in and we kill the connection.

That's too tight - a single dropped packet while waiting for a large result will drop the connection.

If we would like faster detection of downed peers, we should
* set TCP_KEEPINTVL - the interval at which probes are sent - to something smaller, like 5 seconds
* set TCP_KEEPIDLE - the threshold for sending the first probe - to something very small, like 2 seconds
* set TCP_USER_TIMEOUT to allow for a few dropped probes - 25 seconds or so.

Reply:
the 30 seconds comes from here: https://github.com/kubernetes/client-go/blob/master/transport/cache.go#L94

Reply:
sgtm, thx, updated.